src/dist/failure_detector/failure_detector.cpp

```diff
@@ -214,8 +214,13 @@ void failure_detector::process_all_records()
                 }
             }
 
+            /*
+             * "Check interval" and "send beacon" are interleaved, so we must
+             * test if "record will expire before next time we check all the records"
+             * in order to guarantee the perfect fd
+             */
             if (record.is_alive
-                && now - record.last_send_time_for_beacon_with_ack >= _lease_milliseconds)
+                && now + _check_interval_milliseconds - record.last_send_time_for_beacon_with_ack > _lease_milliseconds)
```

Comment:
The figure has an assumption that the time is continuous, but in the code the time is segmented. 
Consider this condition: the check interval is 6 seconds, and let's assume worker starts to check the records in the time tick of 0. So the replica server can detect the master disconnected on the time tick 0, 6, 12, 18, ...6n..., as "process_all_records" are called by a timer. 
Now let's assume the worker send a beacon on time 5, and the lease_period is 14. So ideally, the worker should detect the master expired on time tick of 19. But in the code, the result should be 24, which is 5 seconds later.
As It's not easy to write the test case, this is only my analysis based on the code. So must say sorry if I have misunderstand on the code.